nupic/research/frameworks/vernon/experiments/self_supervised_experiment.py

```diff
@@ -0,0 +1,452 @@
+    def load_dataset(cls, config, train=True, dataset_type="unsupervised"):
+        """
+        Loads one of three types of datasets: unsupervised, supervised, and validation
+
+        The dataset_class argument in the config file should either be a single
+        dataset class, or a dict with the following structure:
+        {
+         "unsupervised": unsupervised_dataset_class
+         "supervised": supervised_dataset_class
+         "validation": validation_dataset_class
+        }
+
+        Similarly, the dataset_args argument in the config should be a dict of dicts
+        with keys corresponding to the different dataset types as described above.
+        """
+        dataset_class = config.get("dataset_class", None)
+        dataset_args = dict(config.get("dataset_args", {}))
+
+        if isinstance(dataset_class, dict):
+            dataset_class = dataset_class.get(dataset_type, None)
+
+        dataset_args = dict(dataset_args.get(dataset_type, {}))
+
+        if dataset_class is None:
+            raise ValueError("Must specify 'dataset_class' in config.")
+
+        return dataset_class(**dataset_args)
+
+    def unsupervised_loss(self, output, target, reduction="mean"):
+        """
+        The error loss component of the loss function.
+        """
+        return self._loss_function_unsupervised(output, reduction=reduction)
+
+    def supervised_loss(self, output, target, reduction="mean"):
+        """
+        The error loss component of the loss function.
+        """
+        return self._loss_function_supervised(output, target, reduction=reduction)
+
+    def encoder_complexity_loss(self, model):
+        """
+        The encoder complexity component of the loss function.
+        """
+        pass
+
+    def pre_batch_supervised(self, model, batch_idx):
+        pass
+
+    def post_batch_supervised(
+        self, model, error_loss, complexity_loss, batch_idx, num_images, time_string
+    ):
+        # Update 1cycle learning rate after every batch
+        if self.step_lr_every_batch_classifier:
+            self.lr_scheduler_classifier.step()
+
+    def post_optimizer_step_supervised(self, model):
+        pass
+
+    def post_batch_wrapper_supervised(self, **kwargs):
+        self.post_optimizer_step_supervised(self.model)
+        self.post_batch_supervised(**kwargs)
+
+    def transform_data_to_device_unsupervised(self, data, target, device, non_blocking):
+        """
+        This provides an extensibility point for performing any final
+        transformations on the data or targets. This will suffice for most
+        self-supervised experiments. The first copy of data gets passed through the
+        model, and the second copy is optionally used in the loss function.
+        """
+        data = data.to(self.device, non_blocking=non_blocking)
+        return data, data
+
+    def train_epoch(self):
+        self.train_model(
+            self.encoder,
+            self.unsupervised_loader,
```

Comment:
Can you please add the parameter names here (and everywhere else you call train/evaluate methods) for ease of readability, especially since these methods have a lot of arguments? This would be just how `SupervisedExperiment`'s train method includes all parameter names when called [here](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/vernon/experiments/supervised_experiment.py#L363-L375).

Reply:
Also, if you haven't already included some documentation for the `train_epoch` and `validate` methods, please do. That would be very helpful since this experiment class has 2 models, so it would help understand what exactly is being trained.

Reply:
Done for both.